### PR TITLE
Add all environment to event listener

### PR DESCRIPTION
**Pull request by Cyperghost**
#2952

- And remove duplicate code

### CODE REVIEW

File: wcfsetup/install/files/lib/system/cache/builder/EventListenerCacheBuilder.class.php [hunk 1]
```diff
@@ -40,32 +40,34 @@ public function rebuild(array $parameters) {
...
-						$inheritedActions[$eventListener->environment][$eventListener->eventClassName] = [];
+					foreach ($eventNames as $eventName) {
+						$key = EventHandler::generateKey($eventListener->eventClassName, $eventName);
+						if (!isset($actions[$environment][$key])) {
+							$actions[$environment][$key] = [];
+						}
+						
+						$actions[$environment][$key][] = $eventListener;
+					}
+				} else {
```
Comment (joshuaruesweg) [inline wcfsetup/install/files/lib/system/cache/builder/EventListenerCacheBuilder.class.php]:
```suggestion
				} 
				else {
```

File: wcfsetup/install/files/lib/system/event/EventHandler.class.php [hunk 2]
```diff
@@ -185,49 +155,83 @@ public function fireAction($eventObj, $eventName, array &$parameters = []) {
...
-				$actionObj->execute($eventObj, $className, $eventName, $parameters);
-				
-				if (!is_array($parameters)) {
-					throw new SystemException("'".get_class($actionObj)."' breaks the '\$parameters' array!");
+			$this->callEvent($eventObj, $actionObj, $className, $eventName, $parameters);
+		}
+	}
+	
+	/**
+	 * Create the event object and check if this implements the interfaces
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/event/EventHandler.class.php:175]:
```suggestion
	 * Validates and creates the event listener.
```

File: wcfsetup/install/files/lib/system/event/EventHandler.class.php [hunk 3]
```diff
@@ -185,49 +155,83 @@ public function fireAction($eventObj, $eventName, array &$parameters = []) {
...
+			
+			return $object;
+		}
+		
+		return null;
+	}
+	
+	/**
+	 * This function calls the event listener
+	 * Function to reduce duplicate code in the fire action between inherited and not inherited actions
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/event/EventHandler.class.php:213]:
```suggestion
	 * Invokes the event listener and checks for unwanted side effects.
```

File: wcfsetup/install/files/lib/system/event/EventHandler.class.php [hunk 4]
```diff
@@ -185,49 +155,83 @@ public function fireAction($eventObj, $eventName, array &$parameters = []) {
...
+	 *
+	 * @param object $eventObj
+	 * @param object $actionObj
+	 * @param string $className
+	 * @param string $eventName
+	 * @param array &$parameters
+	 *
+	 * @throws SystemException
+	 * @throws \BadMethodCallException
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/event/EventHandler.class.php:222]:
```suggestion
```

**Comment by dtdesign** [comment]
@Cyperghost This pull request has been stale for quite some time, please look into the suggested changes and decide on an action.

**Comment by TimWolla** [comment]
I'm only realizing _just now_ that the only functional change related to the title of this PR is the one in the CacheBuilder. The change in the EventHandler is an unrelated optimization.

@Cyperghost Please confirm: Is this realization correct? Then I'd appreciate if you could revert the changes to the EventHandler and propose them in a separate PR. Based off the commit history I expect that we are going to squash the commits during the merge and having two unrelated changes (especially with this large of a refactoring) within a single commit will become messy in the future.

**Comment by Cyperghost** [comment]
@TimWolla yes that’s right. This changes are only optimized code. The changes comming from my other PR.
I‘will remove it.

**Comment by TimWolla** [comment]
Now merged, thanks.

**Pull request merged**